File: v1/haskell/meetup/Meetup.hs
module Meetup (Weekday(..), Schedule(..), meetupDay) where

import Data.Time.Calendar (Day, fromGregorian, gregorianMonthLength)
import Data.Time.Calendar.WeekDate (toWeekDate)

data Weekday = Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday
    deriving (Enum, Eq)

data Schedule = First | Second | Third | Fourth | Last | Teenth
    deriving (Enum, Eq)

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay schedule weekDay year month = 
    case schedule of Teenth -> fst $ head $ take 1 $ filter (\(d,_) -> dayOfMonth d >= 13) weekDays
                     Last   -> fst $ last weekDays
                     x      -> fst $ head $ drop (fromEnum x) weekDays
    where monthDays = daysInMonth year month
          weekDays = filter (\(_,w) -> w == weekDay) monthDays

daysInMonth :: Integer -> Int -> [(Day, Weekday)]
daysInMonth year month = [dateToWeekday year month day | day <- [1..numberOfDays]] where
    numberOfDays = gregorianMonthLength year month

dateToWeekday :: Integer -> Int -> Int -> (Day, Weekday)
dateToWeekday year month day = (gregorianDay, weekDay) where
    gregorianDay = fromGregorian year month day
    (_, _, dayNumber) = toWeekDate gregorianDay
    weekDay = gregorianDayToWeekday dayNumber

dayOfMonth :: Day -> Int
dayOfMonth day = dayNumber where 
    (_, _, dayNumber) = toWeekDate day

gregorianDayToWeekday :: Int -> Weekday
gregorianDayToWeekday day = toEnum (pred day)::Weekday

File: v1/fsharp/meetup/Meetup.fs
using System;
using System.Linq;

public class Meetup
{
    private readonly ILookup<DayOfWeek, DateTime> days;

    public Meetup(int month, int year)
    {
        this.days = Enumerable.Range(1, DateTime.DaysInMonth(year, month))
            .Select(day => new DateTime(year, month, day))
            .ToLookup(d => d.DayOfWeek);
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        var daysOfWeek = this.days[dayOfWeek];

        if (schedule == Schedule.Teenth)
        {
            return daysOfWeek.First(d => d.Day >= 13);
        }

        return daysOfWeek.ElementAt(Math.Min((int)schedule, daysOfWeek.Count() - 1));
    }
}

public enum Schedule
{
    First,
    Second,
    Third,
    Fourth,
    Last,
    Teenth
}